File: mobile/ios/Runner/HuddleMediaPlugin.swift
import AVFoundation
import Flutter
import UIKit

/// Foreground-only native seam for iOS Huddle media.
///
/// Owns microphone permission, the voice-processing audio session, native
/// Opus capture/playout, interruptions, and the built-in output toggle.
final class HuddleMediaPlugin {
  private let channel: FlutterMethodChannel
  private let audioSession = AVAudioSession.sharedInstance()
  private var audioSessionPrepared = false
  private var speakerEnabled = false
  private var audioEngine: HuddleAudioEngine?
  private var interruptionObserver: NSObjectProtocol?
  private var mediaServicesResetObserver: NSObjectProtocol?

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(
      name: "buzz/huddle_media",
      binaryMessenger: messenger
    )
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
    interruptionObserver = NotificationCenter.default.addObserver(
      forName: AVAudioSession.interruptionNotification,
      object: audioSession,
      queue: .main
    ) { [weak self] notification in
      self?.handleInterruption(notification)
    }
    mediaServicesResetObserver = NotificationCenter.default.addObserver(
      forName: AVAudioSession.mediaServicesWereResetNotification,
      object: audioSession,
      queue: .main
    ) { [weak self] _ in
      self?.handleMediaServicesReset()
    }
  }

  deinit {
    channel.setMethodCallHandler(nil)
    audioEngine?.stop()
    audioEngine = nil
    if let interruptionObserver {
      NotificationCenter.default.removeObserver(interruptionObserver)
    }
    if let mediaServicesResetObserver {
      NotificationCenter.default.removeObserver(mediaServicesResetObserver)
    }
    if audioSessionPrepared {
      try? audioSession.overrideOutputAudioPort(.none)
      try? audioSession.setActive(
        false,
        options: [.notifyOthersOnDeactivation]
      )
    }
  }

  private func handle(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    switch call.method {
    case "getCapabilities":
      result(capabilities)
    case "requestMicrophonePermission":
      requestMicrophonePermission(result: result)
    case "openSystemSettings":
      openSystemSettings(result: result)
    case "prepare":
      prepare(arguments: call.arguments, result: result)
    case "start":
      start(result: result)
    case "setMuted":
      setMuted(arguments: call.arguments, result: result)
    case "setSpeakerEnabled":
      setSpeakerEnabled(arguments: call.arguments, result: result)
    case "playRemoteOpusFrame":
      playRemoteOpusFrame(arguments: call.arguments, result: result)
    case "removeRemotePeer":
      removeRemotePeer(arguments: call.arguments, result: result)
    case "stop":
      stop(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private var capabilities: [String: Any] {
    let supportsOpus = HuddleAudioEngine.isSupported()
    return [
      "platform": "ios",
      "audioSession": true,
      "microphonePermission": true,
      "capture": supportsOpus,
      "playback": supportsOpus,
      "opusEncoding": supportsOpus,
      "opusDecoding": supportsOpus,
    ]
  }

  private func requestMicrophonePermission(result: @escaping FlutterResult) {
    switch audioSession.recordPermission {
    case .granted:
      result("granted")
    case .denied:
      result("denied")
    case .undetermined:
      audioSession.requestRecordPermission { granted in
        DispatchQueue.main.async {
          result(granted ? "granted" : "denied")
        }
      }
    @unknown default:
      result("restricted")
    }
  }

  /// Open the iOS Settings app on this app's page so the user can grant a
  /// previously denied microphone permission. iOS never re-prompts once denied,
  /// so this is the only in-app recovery path.
  private func openSystemSettings(result: @escaping FlutterResult) {
    guard let url = URL(string: UIApplication.openSettingsURLString) else {
      result(false)
      return
    }
    DispatchQueue.main.async {
      guard UIApplication.shared.canOpenURL(url) else {
        result(false)
        return
      }
      UIApplication.shared.open(url, options: [:]) { opened in
        result(opened)
      }
    }
  }

  private func prepare(arguments: Any?, result: @escaping FlutterResult) {
    guard let values = arguments as? [String: Any],
      (values["protocolVersion"] as? NSNumber)?.intValue == 3,
      (values["sampleRateHz"] as? NSNumber)?.intValue == 48_000,
      (values["channels"] as? NSNumber)?.intValue == 1,
      (values["frameSamples"] as? NSNumber)?.intValue == 960
    else {
      result(
        FlutterError(
          code: "invalid_configuration",
          message: "Expected the fixed Huddle Opus v3 media configuration.",
          details: nil
        )
      )
      return
    }
    guard HuddleAudioEngine.isSupported() else {
      result(
        FlutterError(
          code: "unsupported",
          message: "This iOS device does not expose native Opus encode/decode.",
          details: nil
        )
      )
      return
    }
    guard audioSession.recordPermission == .granted else {
      result(
        FlutterError(
          code: "microphone_permission_denied",
          message: "Microphone permission is required for a Huddle.",
          details: nil
        )
      )
      return
    }

    do {
      try audioSession.setCategory(
        .playAndRecord,
        mode: .voiceChat,
        options: [.allowBluetoothHFP]
      )
      try audioSession.setPreferredSampleRate(48_000)
      try audioSession.setPreferredIOBufferDuration(0.02)
      try audioSession.setActive(true)
      try audioSession.overrideOutputAudioPort(.none)
      audioSessionPrepared = true
      speakerEnabled = false
      result([
        "audioSessionPrepared": true,
        "sampleRateHz": Int(audioSession.sampleRate),
        "channels": 1,
        "frameSamples": 960,
      ])
    } catch {
      audioSessionPrepared = false
      try? audioSession.overrideOutputAudioPort(.none)
      try? audioSession.setActive(
        false,
        options: [.notifyOthersOnDeactivation]
      )
      result(
        FlutterError(
          code: "audio_session_failed",
          message: "Unable to prepare the iOS Huddle audio session.",
          details: error.localizedDescription
        )
      )
    }
  }

  private func start(result: @escaping FlutterResult) {
    guard audioSessionPrepared else {
      result(
        FlutterError(
          code: "invalid_state",
          message: "Prepare Huddle audio before starting it.",
          details: nil
        )
      )
      return
    }
    guard audioEngine == nil else {
      result(
        FlutterError(
          code: "invalid_state",
          message: "Huddle audio is already running.",
          details: nil
        )
      )
      return
    }

    do {
      let engine = try HuddleAudioEngine(
        onLocalPacket: { [weak self] packet in
          self?.emitLocalPacket(packet)
        },
        onFailure: { [weak self] code, message in
          self?.emitNativeFailure(code: code, message: message)
        },
        onDiagnostics: { [weak self] diagnostics in
          self?.emitCaptureDiagnostics(diagnostics)
        },
        diagnosticsEnabled: Self.diagnosticsEnabled
      )
      try engine.start()
      audioEngine = engine
      result(nil)
    } catch {
      result(
        FlutterError(
          code: "media_start_failed",
          message: "Unable to start iOS Huddle audio.",
          details: error.localizedDescription
        )
      )
    }
  }

  private func setMuted(arguments: Any?, result: @escaping FlutterResult) {
    guard let values = arguments as? [String: Any],
      let muted = values["muted"] as? Bool
    else {
      result(
        FlutterError(
          code: "invalid_arguments",
          message: "Missing Huddle mute state.",
          details: nil
        )
      )
      return
    }
    do {
      guard let audioEngine else {
        throw HuddleNativeMediaError.invalidState(
          "Huddle audio is not running."
        )
      }
      try audioEngine.setMuted(muted)
      result(nil)
    } catch {
      result(
        FlutterError(
          code: "invalid_state",
          message: error.localizedDescription,
          details: nil
        )
      )
    }
  }

  private func setSpeakerEnabled(
    arguments: Any?,
    result: @escaping FlutterResult
  ) {
    guard audioSessionPrepared, audioEngine != nil else {
      result(
        FlutterError(
          code: "invalid_state",
          message: "Huddle audio is not running.",
          details: nil
        )
      )
      return
    }
    guard let values = arguments as? [String: Any],
      let enabled = values["enabled"] as? Bool
    else {
      result(
        FlutterError(
          code: "invalid_arguments",
          message: "Missing Huddle speaker state.",
          details: nil
        )
      )
      return
    }

    do {
      try audioSession.overrideOutputAudioPort(enabled ? .speaker : .none)
      speakerEnabled = enabled
      result(nil)
    } catch {
      result(
        FlutterError(
          code: "audio_route_failed",
          message: "Unable to change Huddle audio output.",
          details: error.localizedDescription
        )
      )
    }
  }

  private func playRemoteOpusFrame(
    arguments: Any?,
    result: @escaping FlutterResult
  ) {
    guard let values = arguments as? [String: Any],
      let peerIndex = (values["peerIndex"] as? NSNumber)?.intValue,
      (0...255).contains(peerIndex),
      let sequence = (values["sequence"] as? NSNumber)?.intValue,
      (0...0xffff).contains(sequence),
      let timestamp = (values["timestamp48k"] as? NSNumber)?.int64Value,
      (0...Int64(UInt32.max)).contains(timestamp),
      let levelDbov = (values["levelDbov"] as? NSNumber)?.intValue,
      (-127...0).contains(levelDbov),
      let typedData = values["opus"] as? FlutterStandardTypedData,
      !typedData.data.isEmpty,
      typedData.data.count <= HuddleAudioFormats.maximumOpusPacketBytes
    else {
      result(
        FlutterError(
          code: "invalid_arguments",
          message: "Malformed remote Huddle Opus packet.",
          details: nil
        )
      )
      return
    }

    do {
      guard let audioEngine else {
        throw HuddleNativeMediaError.invalidState(
          "Huddle audio is not running."
        )
      }
      try audioEngine.enqueueRemote(
        HuddleRemoteOpusPacket(
          peerIndex: peerIndex,
          sequence: sequence,
          timestamp48k: timestamp,
          levelDbov: levelDbov,
          opus: typedData.data
        )
      )
      result(nil)
    } catch {
      result(
        FlutterError(
          code: "playback_failed",
          message: error.localizedDescription,
          details: nil
        )
      )
    }
  }

  private func removeRemotePeer(
    arguments: Any?,
    result: @escaping FlutterResult
  ) {
    guard let values = arguments as? [String: Any],
      let peerIndex = (values["peerIndex"] as? NSNumber)?.intValue,
      (0...255).contains(peerIndex)
    else {
      result(
        FlutterError(
          code: "invalid_arguments",
          message: "Missing Huddle peer index.",
          details: nil
        )
      )
      return
    }
    guard let audioEngine else {
      result(
        FlutterError(
          code: "invalid_state",
          message: "Huddle audio is not running.",
          details: nil
        )
      )
      return
    }
    audioEngine.removeRemotePeer(peerIndex)
    result(nil)
  }

  private func handleInterruption(_ notification: Notification) {
    guard audioSessionPrepared,
      let rawType = notification.userInfo?[AVAudioSessionInterruptionTypeKey]
        as? NSNumber,
      let type = AVAudioSession.InterruptionType(rawValue: rawType.uintValue)
    else { return }

    if type == .began {
      audioEngine?.setInterrupted(true)
      emitInterruptionChanged(true)
      return
    }
    let rawOptions =
      notification.userInfo?[AVAudioSessionInterruptionOptionKey]
      as? NSNumber
    let options = AVAudioSession.InterruptionOptions(
      rawValue: rawOptions?.uintValue ?? 0
    )
    guard options.contains(.shouldResume) else { return }
    do {
      try audioSession.setActive(true)
      if speakerEnabled {
        try audioSession.overrideOutputAudioPort(.speaker)
      }
      audioEngine?.setInterrupted(false)
      emitInterruptionChanged(false)
    } catch {
      emitNativeFailure(
        code: "audio_resume_failed",
        message: "Unable to resume iOS Huddle audio after interruption."
      )
    }
  }

  private func handleMediaServicesReset() {
    guard audioSessionPrepared || audioEngine != nil else { return }
    audioEngine?.stop()
    audioEngine = nil
    audioSessionPrepared = false
    speakerEnabled = false
    emitNativeFailure(
      code: "media_services_reset",
      message: "iOS audio services restarted. Rejoin the Huddle to continue."
    )
  }

  private func stop(result: @escaping FlutterResult) {
    audioEngine?.stop()
    audioEngine = nil
    guard audioSessionPrepared else {
      speakerEnabled = false
      result(nil)
      return
    }
    do {
      try audioSession.overrideOutputAudioPort(.none)
      try audioSession.setActive(
        false,
        options: [.notifyOthersOnDeactivation]
      )
      audioSessionPrepared = false
      speakerEnabled = false
      result(nil)
    } catch {
      result(
        FlutterError(
          code: "audio_session_stop_failed",
          message: "Unable to stop the iOS Huddle audio session.",
          details: error.localizedDescription
        )
      )
    }
  }

  private func emitLocalPacket(_ packet: HuddleLocalOpusPacket) {
    DispatchQueue.main.async { [weak self] in
      guard self?.audioEngine != nil else { return }
      self?.channel.invokeMethod(
        "localOpusFrame",
        arguments: [
          "sequence": packet.sequence,
          "timestamp48k": packet.timestamp48k,
          "levelDbov": packet.levelDbov,
          "flags": packet.flags,
          "opus": FlutterStandardTypedData(bytes: packet.opus),
        ]
      )
    }
  }

  private func emitNativeFailure(code: String, message: String) {
    DispatchQueue.main.async { [weak self] in
      guard self?.audioEngine != nil || code == "media_services_reset" else {
        return
      }
      self?.channel.invokeMethod(
        "nativeError",
        arguments: ["code": code, "message": message]
      )
    }
  }

  private func emitCaptureDiagnostics(
    _ diagnostics: HuddleCaptureDiagnostics
  ) {
    DispatchQueue.main.async { [weak self] in
      guard self?.audioEngine != nil else { return }
      self?.channel.invokeMethod(
        "captureDiagnostics",
        arguments: [
          "frameCount": diagnostics.frameCount,
          "rmsDbovHistogram": diagnostics.rmsDbovHistogram,
          "peakDbovHistogram": diagnostics.peakDbovHistogram,
          "maxPeakDbov": diagnostics.maxPeakDbov,
          "audioSource": 0,
          "audioSourceLabel": "voice_processing_io",
          "deviceId": nil,
          "deviceType": nil,
          "deviceLabel": diagnostics.deviceLabel,
          "acousticEchoCancelerAvailable": true,
          "acousticEchoCancelerEnabled": diagnostics.voiceProcessingEnabled,
          "noiseSuppressorAvailable": false,
          "noiseSuppressorEnabled": nil,
          "automaticGainControlAvailable": false,
          "automaticGainControlEnabled": nil,
        ]
      )
    }
  }

  private func emitInterruptionChanged(_ interrupted: Bool) {
    channel.invokeMethod(
      "interruptionChanged",
      arguments: ["interrupted": interrupted]
    )
  }

  private static var diagnosticsEnabled: Bool {
    #if DEBUG
      true
    #else
      false
    #endif
  }
}

File: mobile/android/app/src/main/kotlin/xyz/block/buzz/mobile/HuddleMediaPlugin.kt
package xyz.block.buzz.mobile

import android.Manifest
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Foreground-only native seam for mobile Huddle media.
 *
 * Owns microphone permission, foreground communication routing, and the native
 * realtime Opus engine. PCM remains native; Flutter only handles compressed
 * Huddle v3 packets.
 */
internal class HuddleMediaPlugin(
    private val activity: Activity,
    messenger: BinaryMessenger,
) {
    private val channel = MethodChannel(messenger, CHANNEL_NAME)
    private val audioManager = activity.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private val mainHandler = Handler(Looper.getMainLooper())
    private var previousAudioMode: Int? = null
    private var previousCommunicationDevice: AudioDeviceInfo? = null
    private var selectedCommunicationDevice = false
    private var previousSpeakerphoneOn: Boolean? = null
    private var pendingPermissionResult: MethodChannel.Result? = null
    private var audioEngine: HuddleAudioEngine? = null
    private var hasAudioFocus = false
    private val audioFocusListener = AudioManager.OnAudioFocusChangeListener { change ->
        val isInterrupted = change != AudioManager.AUDIOFOCUS_GAIN
        audioEngine?.setInterrupted(isInterrupted)
        mainHandler.post {
            if (audioEngine != null) {
                channel.invokeMethod(
                    "interruptionChanged",
                    mapOf("interrupted" to isInterrupted),
                )
            }
        }
    }

    init {
        channel.setMethodCallHandler(::handleMethodCall)
    }

    private fun handleMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            "getCapabilities" -> result.success(capabilities())
            "requestMicrophonePermission" -> requestMicrophonePermission(result)
            "openSystemSettings" -> openSystemSettings(result)
            "prepare" -> prepare(call.arguments, result)
            "start" -> start(result)
            "setMuted" -> setMuted(call.arguments, result)
            "setSpeakerEnabled" -> setSpeakerEnabled(call.arguments, result)
            "playRemoteOpusFrame" -> playRemoteOpusFrame(call.arguments, result)
            "removeRemotePeer" -> removeRemotePeer(call.arguments, result)
            "stop" -> stop(result)
            else -> result.notImplemented()
        }
    }

    private fun capabilities(): Map<String, Any> {
        val supportsOpus = HuddleAudioEngine.isSupported()
        return mapOf(
            "platform" to "android",
            "audioSession" to true,
            "microphonePermission" to true,
            "capture" to supportsOpus,
            "playback" to supportsOpus,
            "opusEncoding" to supportsOpus,
            "opusDecoding" to supportsOpus,
        )
    }

    private fun requestMicrophonePermission(result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M || hasMicrophonePermission()) {
            result.success("granted")
            return
        }
        if (pendingPermissionResult != null) {
            result.error(
                "permission_request_in_progress",
                "A microphone permission request is already active.",
                null,
            )
            return
        }
        pendingPermissionResult = result
        activity.requestPermissions(
            arrayOf(Manifest.permission.RECORD_AUDIO),
            MICROPHONE_PERMISSION_REQUEST,
        )
    }

    fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ): Boolean {
        if (requestCode != MICROPHONE_PERMISSION_REQUEST) return false
        val result = pendingPermissionResult ?: return true
        pendingPermissionResult = null
        val microphoneIndex = permissions.indexOf(Manifest.permission.RECORD_AUDIO)
        val granted = microphoneIndex >= 0 &&
            grantResults.getOrNull(microphoneIndex) == PackageManager.PERMISSION_GRANTED
        result.success(if (granted) "granted" else "denied")
        return true
    }

    /**
     * Open this app's Android settings screen so the user can grant a
     * previously denied microphone permission. Mirrors the iOS recovery path.
     */
    private fun openSystemSettings(result: MethodChannel.Result) {
        val intent = Intent(
            Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
            Uri.fromParts("package", activity.packageName, null),
        ).apply { addFlags(Intent.FLAG_ACTIVITY_NEW_TASK) }
        try {
            activity.startActivity(intent)
            result.success(true)
        } catch (error: ActivityNotFoundException) {
            result.success(false)
        }
    }

    private fun prepare(arguments: Any?, result: MethodChannel.Result) {
        val values = arguments as? Map<*, *>
        if (values?.get("protocolVersion") != 3 ||
            values["sampleRateHz"] != 48_000 ||
            values["channels"] != 1 ||
            values["frameSamples"] != 960
        ) {
            result.error(
                "invalid_configuration",
                "Expected the fixed Huddle Opus v3 media configuration.",
                null,
            )
            return
        }
        if (!hasMicrophonePermission()) {
            result.error(
                "microphone_permission_denied",
                "Microphone permission is required for a Huddle.",
                null,
            )
            return
        }
        if (!HuddleAudioEngine.isSupported()) {
            result.error(
                "unsupported",
                "This Android device does not expose Opus encode/decode codecs.",
                null,
            )
            return
        }

        try {
            if (previousAudioMode == null) previousAudioMode = audioManager.mode
            audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
            @Suppress("DEPRECATION")
            val focusResult = audioManager.requestAudioFocus(
                audioFocusListener,
                AudioManager.STREAM_VOICE_CALL,
                AudioManager.AUDIOFOCUS_GAIN_TRANSIENT,
            )
            check(focusResult == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                "Android did not grant communication audio focus."
            }
            hasAudioFocus = true
            selectForegroundOutput()
            result.success(
                mapOf(
                    "audioSessionPrepared" to true,
                    "sampleRateHz" to 48_000,
                    "channels" to 1,
                    "frameSamples" to 960,
                ),
            )
        } catch (error: RuntimeException) {
            restoreAudioSession()
            result.error(
                "audio_session_failed",
                "Unable to prepare the Android Huddle audio session.",
                error.message,
            )
        }
    }

    private fun start(result: MethodChannel.Result) {
        if (previousAudioMode == null) {
            result.error(
                "invalid_state",
                "Prepare Huddle audio before starting it.",
                null,
            )
            return
        }
        if (audioEngine != null) {
            result.error("invalid_state", "Huddle audio is already running.", null)
            return
        }

        try {
            val engine = HuddleAudioEngine(
                onLocalPacket = ::emitLocalPacket,
                onFailure = ::emitNativeFailure,
                onDiagnostics = ::emitCaptureDiagnostics,
                diagnosticsEnabled =
                    (activity.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0,
            )
            engine.start()
            audioEngine = engine
            result.success(null)
        } catch (error: Throwable) {
            result.error(
                "media_start_failed",
                "Unable to start Android Huddle audio.",
                error.message,
            )
        }
    }

    private fun setMuted(arguments: Any?, result: MethodChannel.Result) {
        val muted = (arguments as? Map<*, *>)?.get("muted") as? Boolean
        if (muted == null) {
            result.error("invalid_arguments", "Missing Huddle mute state.", null)
            return
        }
        try {
            val engine = audioEngine ?: error("Huddle audio is not running.")
            engine.setMuted(muted)
            result.success(null)
        } catch (error: Throwable) {
            result.error("invalid_state", error.message, null)
        }
    }

    private fun setSpeakerEnabled(arguments: Any?, result: MethodChannel.Result) {
        val enabled = (arguments as? Map<*, *>)?.get("enabled") as? Boolean
        if (enabled == null) {
            result.error("invalid_arguments", "Missing Huddle speaker state.", null)
            return
        }
        if (previousAudioMode == null || audioEngine == null) {
            result.error("invalid_state", "Huddle audio is not running.", null)
            return
        }

        try {
            routeToBuiltInOutput(enabled)
            result.success(null)
        } catch (error: RuntimeException) {
            result.error("audio_route_failed", "Unable to change Huddle audio output.", error.message)
        }
    }

    private fun playRemoteOpusFrame(arguments: Any?, result: MethodChannel.Result) {
        val values = arguments as? Map<*, *>
        val peerIndex = (values?.get("peerIndex") as? Number)?.toInt()
        val sequence = (values?.get("sequence") as? Number)?.toInt()
        val timestamp48k = (values?.get("timestamp48k") as? Number)?.toLong()
        val levelDbov = (values?.get("levelDbov") as? Number)?.toInt()
        val opus = values?.get("opus") as? ByteArray
        if (peerIndex == null || peerIndex !in 0..255 ||
            sequence == null || sequence !in 0..0xffff ||
            timestamp48k == null || timestamp48k !in 0..0xffff_ffffL ||
            levelDbov == null || levelDbov !in -127..0 ||
            opus == null || opus.isEmpty() || opus.size > MAX_OPUS_PACKET_BYTES
        ) {
            result.error("invalid_arguments", "Malformed remote Huddle Opus packet.", null)
            return
        }

        try {
            val engine = audioEngine ?: error("Huddle audio is not running.")
            engine.enqueueRemote(
                HuddleRemoteOpusPacket(
                    peerIndex = peerIndex,
                    sequence = sequence,
                    timestamp48k = timestamp48k,
                    levelDbov = levelDbov,
                    opus = opus.copyOf(),
                ),
            )
            result.success(null)
        } catch (error: Throwable) {
            result.error("playback_failed", error.message, null)
        }
    }

    private fun removeRemotePeer(arguments: Any?, result: MethodChannel.Result) {
        val peerIndex = ((arguments as? Map<*, *>)?.get("peerIndex") as? Number)?.toInt()
        if (peerIndex == null || peerIndex !in 0..255) {
            result.error("invalid_arguments", "Missing Huddle peer index.", null)
            return
        }
        try {
            val engine = audioEngine ?: error("Huddle audio is not running.")
            engine.removeRemotePeer(peerIndex)
            result.success(null)
        } catch (error: Throwable) {
            result.error("playback_failed", error.message, null)
        }
    }

    private fun stop(result: MethodChannel.Result) {
        try {
            audioEngine?.stop()
            audioEngine = null
            restoreAudioSession()
            result.success(null)
        } catch (error: RuntimeException) {
            result.error(
                "audio_session_stop_failed",
                "Unable to stop the Android Huddle audio session.",
                error.message,
            )
        }
    }

    fun dispose() {
        runCatching { audioEngine?.stop() }
        audioEngine = null
        restoreAudioSession()
        pendingPermissionResult?.error(
            "disposed",
            "Huddle media was disposed during microphone permission.",
            null,
        )
        pendingPermissionResult = null
        channel.setMethodCallHandler(null)
    }

    private fun emitLocalPacket(packet: HuddleLocalOpusPacket) {
        mainHandler.post {
            if (audioEngine == null) return@post
            channel.invokeMethod(
                "localOpusFrame",
                mapOf(
                    "sequence" to packet.sequence,
                    "timestamp48k" to packet.timestamp48k,
                    "levelDbov" to packet.levelDbov,
                    "flags" to packet.flags,
                    "opus" to packet.opus,
                ),
            )
        }
    }

    private fun emitNativeFailure(code: String, message: String) {
        mainHandler.post {
            if (audioEngine == null) return@post
            channel.invokeMethod(
                "nativeError",
                mapOf("code" to code, "message" to message),
            )
        }
    }

    private fun emitCaptureDiagnostics(diagnostics: HuddleCaptureDiagnostics) {
        mainHandler.post {
            if (audioEngine == null) return@post
            channel.invokeMethod(
                "captureDiagnostics",
                mapOf(
                    "frameCount" to diagnostics.frameCount,
                    "rmsDbovHistogram" to diagnostics.rmsDbovHistogram,
                    "peakDbovHistogram" to diagnostics.peakDbovHistogram,
                    "maxPeakDbov" to diagnostics.maxPeakDbov,
                    "audioSource" to diagnostics.audioSource,
                    "audioSourceLabel" to diagnostics.audioSourceLabel,
                    "deviceId" to diagnostics.deviceId,
                    "deviceType" to diagnostics.deviceType,
                    "deviceLabel" to diagnostics.deviceLabel,
                    "acousticEchoCancelerAvailable" to
                        diagnostics.acousticEchoCancelerAvailable,
                    "acousticEchoCancelerEnabled" to
                        diagnostics.acousticEchoCancelerEnabled,
                    "noiseSuppressorAvailable" to diagnostics.noiseSuppressorAvailable,
                    "noiseSuppressorEnabled" to diagnostics.noiseSuppressorEnabled,
                    "automaticGainControlAvailable" to
                        diagnostics.automaticGainControlAvailable,
                    "automaticGainControlEnabled" to
                        diagnostics.automaticGainControlEnabled,
                ),
            )
        }
    }

    @Suppress("DEPRECATION")
    private fun selectForegroundOutput() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (selectedCommunicationDevice) return
            previousCommunicationDevice = audioManager.communicationDevice
            val currentType = previousCommunicationDevice?.type
            if (currentType in externalCommunicationDeviceTypes) return
            val earpiece = audioManager.availableCommunicationDevices.firstOrNull {
                it.type == AudioDeviceInfo.TYPE_BUILTIN_EARPIECE
            } ?: return
            selectedCommunicationDevice = audioManager.setCommunicationDevice(earpiece)
            return
        }
        if (previousSpeakerphoneOn == null) {
            previousSpeakerphoneOn = audioManager.isSpeakerphoneOn
            audioManager.isSpeakerphoneOn = false
        }
    }

    @Suppress("DEPRECATION")
    private fun routeToBuiltInOutput(useSpeaker: Boolean) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (previousCommunicationDevice == null) {
                previousCommunicationDevice = audioManager.communicationDevice
            }
            val targetType = if (useSpeaker) {
                AudioDeviceInfo.TYPE_BUILTIN_SPEAKER
            } else {
                AudioDeviceInfo.TYPE_BUILTIN_EARPIECE
            }
            val target = audioManager.availableCommunicationDevices.firstOrNull {
                it.type == targetType
            } ?: error("The requested built-in Huddle output is unavailable.")
            check(audioManager.setCommunicationDevice(target)) {
                "Android rejected the requested Huddle output."
            }
            selectedCommunicationDevice = true
            return
        }

        if (previousSpeakerphoneOn == null) {
            previousSpeakerphoneOn = audioManager.isSpeakerphoneOn
        }
        audioManager.isSpeakerphoneOn = useSpeaker
    }

    @Suppress("DEPRECATION")
    private fun restoreForegroundOutput() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (selectedCommunicationDevice) {
                val previous = previousCommunicationDevice
                if (previous != null &&
                    audioManager.availableCommunicationDevices.any { it.id == previous.id }
                ) {
                    audioManager.setCommunicationDevice(previous)
                } else {
                    audioManager.clearCommunicationDevice()
                }
            }
            previousCommunicationDevice = null
            selectedCommunicationDevice = false
            return
        }
        previousSpeakerphoneOn?.let { audioManager.isSpeakerphoneOn = it }
        previousSpeakerphoneOn = null
    }

    private fun restoreAudioSession() {
        if (hasAudioFocus) {
            @Suppress("DEPRECATION")
            audioManager.abandonAudioFocus(audioFocusListener)
            hasAudioFocus = false
        }
        runCatching { restoreForegroundOutput() }
        runCatching {
            previousAudioMode?.let { audioManager.mode = it }
        }
        previousAudioMode = null
    }

    private fun hasMicrophonePermission(): Boolean =
        Build.VERSION.SDK_INT < Build.VERSION_CODES.M ||
            activity.checkSelfPermission(Manifest.permission.RECORD_AUDIO) ==
            PackageManager.PERMISSION_GRANTED

    private companion object {
        const val CHANNEL_NAME = "buzz/huddle_media"
        const val MICROPHONE_PERMISSION_REQUEST = 7342
        const val MAX_OPUS_PACKET_BYTES = 4088

        val externalCommunicationDeviceTypes = setOf(
            AudioDeviceInfo.TYPE_BLUETOOTH_A2DP,
            AudioDeviceInfo.TYPE_BLUETOOTH_SCO,
            AudioDeviceInfo.TYPE_BLE_HEADSET,
            AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
            AudioDeviceInfo.TYPE_WIRED_HEADSET,
            AudioDeviceInfo.TYPE_USB_DEVICE,
            AudioDeviceInfo.TYPE_USB_HEADSET,
        )
    }
}
